azerothcore/azerothcore-wotlk
# Fix improve our docker structure

## Changes Proposed:
- use Dockerfile multi-stage feature to avoid having a Dockerfile for "build" and another for "run"
- merge worldserver and authserver images since they use the same codebase. Introduce a flag START_AS_AUTHSERVER to be able to run the image as authserver
- use Docker BuildKit in order to enable layers caching in pipelines
- enable compilation cache when building with Docker
- publish the images of containers on docker hub
- allowing the user to pull and use them without necessarily building them locally


## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/4006



## Tests Performed:
- `COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 docker-compose up --build` everything seems to be up and running
- change a C++ file, run `COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 docker-compose build` and observe that compilation is faster than the first time
- run pipeline, wait the build to finish, then force rerun by pushing an empty commit `git commit -m "Force rerun" --allow-empty` and observe that layers are cached 



## How to Test the Changes:

- `COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 docker-compose up --build`
- change a C++ file, run `COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 docker-compose build`

## Known Issues and TODO List:

- add this to documentation: In order to build a testable image and run unit tests, do the following: `DOCKER_BUILDKIT=1 docker build -f docker/server/Dockerfile --target=testable -t azerothcore2/server:test .` then `docker run azerothcore2/server:test ./azerothcore/build/src/test/unit_tests`
- when runnung tests with the method described above there very few tests run. Is it expected ?


## Target Branch(es):








 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
